From 537f31bd150b7130df1f49d37f5e31788366c129 Mon Sep 17 00:00:00 2001
From: Noam Camus <ncamus@habana.ai>
Date: Tue, 16 Apr 2019 16:55:13 +0300
Subject: [PATCH] [SW-0] spi: update BAUDR

Update baudr register and do not count on BTL value
This is true since we may change PLL which SPI depends on (PCI)

Change-Id: I53916b14d3590bcde1526582f340db7616fce698
---

diff --git a/common/preboot.h b/common/preboot.h
index b7515cf..ecc353e 100644
--- a/common/preboot.h
+++ b/common/preboot.h
@@ -46,5 +46,6 @@
 
 u64 get_cpu_boot_status_reg(void);
 void hl_uart_init(void);
+u32 hl_spi_baudr_get(void);
 
 #endif /* COMMON_PREBOOT_H_ */
diff --git a/common/spi.c b/common/spi.c
index 650d3c7..c979432 100644
--- a/common/spi.c
+++ b/common/spi.c
@@ -192,9 +192,14 @@
 
 static void spi_init(void)
 {
+	u32 baudr = hl_spi_baudr_get();
+
+	/* Validate that baudr is even */
+	baudr = (baudr + 1) & ~1;
+
 	spi_ssi_cs0_init();
 	regmap_writel(0x00000000, mmPSOC_SPI_SSIENR);
-	regmap_writel(0x00000002, mmPSOC_SPI_BAUDR);	/* Min divisor */
+	regmap_writel(baudr, mmPSOC_SPI_BAUDR);
 	regmap_writel(0x00000000, mmPSOC_SPI_RXFTLR);
 	regmap_writel(0x00000000, mmPSOC_SPI_TXFTLR);
 	regmap_writel(0x0000003F, mmPSOC_SPI_IMR);	/* Disable Interrupts */
diff --git a/gaudi/gaudi.c b/gaudi/gaudi.c
index 1eb4088..646ff84 100644
--- a/gaudi/gaudi.c
+++ b/gaudi/gaudi.c
@@ -59,3 +59,8 @@
 {
 	preboot_uart_init(UART_BAUDDIV);
 }
+
+u32 hl_spi_baudr_get(void)
+{
+	return SPI_BAUDR;
+}
diff --git a/gaudi/gaudi.h b/gaudi/gaudi.h
index e2782d6..a23cdeb 100644
--- a/gaudi/gaudi.h
+++ b/gaudi/gaudi.h
@@ -14,6 +14,7 @@
 
 #define PCI_DIV2_CLOCK			(830000000 / 5)
 #define UART_BAUDDIV			(PCI_DIV2_CLOCK / 115200)
+#define SPI_BAUDR			(PCI_DIV2_CLOCK / 25000000)
 
 void preboot_pcie_wait_for_link(void);
 static inline void preboot_delay(void) {}
diff --git a/goya/goya.c b/goya/goya.c
index afa1dcb..fde8fd9 100644
--- a/goya/goya.c
+++ b/goya/goya.c
@@ -191,3 +191,7 @@
 
 void hl_uart_init(void) {};
 
+u32 hl_spi_baudr_get(void)
+{
+	return 0x2;	/* Min divisor */
+}
